### ISSUE TITLE
Feature to group jobs with snakemake

### ISSUE THREAD

**Issue by rahulbshrestha** [marrlab/DomainLab#121]
- Is there is possibility to group jobs together so that the can be run separately and aggregated later?
- For parallel jobs:  
       - different random seed
       - different hyperparameters for the same algorithm 
       - different algorithms
- If resources in the GPU aren't enough to support parallel jobs, will a single job survive or will all jobs fail?
- Is there a mechanism to rerun failed jobs if the failure wasn't caused by a software bug? Does this failed job have an "id" so that it can be run again with snakemake?

**Comment by schoersch** [comment]
regarding 2. by the current design, each jobs contains one experiment, which means in this case that:

1. algorithm and hyperparameter is fixed.
2. for this combination, it iterates over all leave-one-out domains and all random seeds.

So parallelizing below the per hyperparameter sample choice and algorithm would need to break up this unit.

In other words:
we have `number of hyperparmeter samples * number of tasks` jobs, and each job trains its algorithm for `number of leave-out-domains * number of seeds` times.

**Comment by schoersch** [comment]
@Car-la-F pointed out, that we could split the computation between different cluster rather easily, if we run the same configuration except different random seeds. By this, we can split the repetition for stochastic variance.

**Comment by Car-la-F** [comment]
Actually I think it would be easier to split the computation over different hyperparameter samples, by using different random seeds. Hence we are not restricted to the performance of the slower VM

**Comment by schoersch** [comment]
@smilesun @rahulbshrestha @xinyuejohn 

I did some tests on the denbi vm and calculated with @Car-la-F the following:

1. It seems reasonable to reduce the image size to 48x48 pixels, although this is indeed small.
2. Considered are 5 algorithms: deepall, dann, deepall_dial, diva, hduva. matchdg is excluded due to #126 , jigen since it seems not stable yet, #129 .
3. We tests against each domain as leave-one-out
4. We consider only 2 random seeds per leave-out-domain (`num_seeds`) -> leading to 6 datapoints per hyperparameter sample.
5. bs=64
6. with this setup, the denbi server can run 5 processes in parallel (`cores`)

By this, we observe the following:
1. loading the dataset to memory takes ~160s (`load_dset`)
2. train all 5 algos for 1 epoch takes ~1365s (`per_epoch`)

Computing budget:
- we can compute for about 4 days if we start by **tonight** (`days`)
- So we want to set the number of epos and hyperparameter samples accordingly.

Based on the numbers above, we obtain:
`hsamples(epos) = (days * 24 * 60 * 60 * cores/num_seeds) / (3 * per_epoch * epos + 3 * load_dset * 5)`
which gives for the denbi vm:

__hsamples=2 hyperparameter samples__ with __epos=100__

or

__hsamples=4 hyperparameter samples__ with __epos=100__ using __only 1 random seed__.

**Comment by Car-la-F** [comment]
@rahulbshrestha @xinyuejohn,

Georg and I agreed on using just 1 random seed, as already 4 hyperparameter samples feel quite less. If you think we should use 2 random seeds, we can still change it. 
Hopefully we can get some more hyperparameter samples from the helmholz vm.


It would be good if you could do the same evaluations as we did above for the helmholz cluster. To be sure that you use the same setup as we did I cloned the branch we were working on to create a identical branch for you: `helm_benchmark`. 
On this branch you will find two yaml files for the helmholz cluster:
1. `helm_runtime_evaluation.yaml`
2. `test_helm_benchmark.yaml`

**Runtime evaluation:**
You can use yaml file 1. for evaluating the runtime with the same setup we used (which means you can use the formula from above). Doing so you need to: 
- edit `run_snakemake.sh` to call the helmholz config instead of the denbi config 
- set cores -> `cores`

For using the formular:
- note down the time for loading the dataset to memory -> `load_dset` [s]
- note down the final time for each algorithm and devide by 2 (we use 2 epochs in the config) -> `per_epoch` [s]
- use `num_seeds` = 1, `days` = 4 (if you can start the benchmark today), `epos` = 100

**Benchmark:**
For the final benchmark you can use yaml file 2.
- set `num_param_samples` according to your results from the formular
- set cores in `run_snakemake.sh`

**Comment by schoersch** [comment]
@rahulbshrestha @xinyuejohn @smilesun 

To clarify:
Currently the helm branch samples different hyperparameters and uses a different random seed for model initialization and training.

**Comment by schoersch** [comment]
@smilesun 
@xinyuejohn set a `RandomResizedCrop` transform, I guess for data augmentation.
In my understanding, using `DsetInMemDecorator` applies the transformations only once, when the dataset is collected in the list. Thus, I'm about the disable the random crop, as for our data it holds true that the cell is centered in the image and the dataset variance is not increased by the random crop.

**Comment by schoersch** [comment]
@smilesun @rahulbshrestha @xinyuejohn @Car-la-F 
I started the benchmark on denbi with 4 hyperparameter samples, epos=100 and 1 random seed.

To make the code of the benchmark easy accessible, I created the `bloodcell_benchmark_denbi` tag, which points to the code running in the benchmark.

**Comment by Car-la-F** [comment]
@rahulbshrestha @xinyuejohn @smilesun 

@schoersch and I checked the progress on the denbi vm today. 
It seems like we did make a mistake when evaluating the time needed for training. We did start all 5 algorithms in 5 cores in parallel. After the deepall and dann finished they released a lot of computing power, hence we underestimated the time for hduva, diva and dial.

@schoersch stopped the benchmark and we do a second evaluation of the runtime using just one core. This one core does then use ~60-80% of the resources. In the benchmark we plan to use `cores` = 4 (as this will use the full gpu), by using the estimation we made with one core 
-> we will more likely overestimate the runtime

We also reduced the number of epos to `epos` = 50. This is motivated by the results of deepall, the only algorithm which could finish one task since yesterday. The result of this algorithm is not better than the one we received for the 2 epos.

**Our new Benchmark settings:**
- `epos` = 50 (will reduce runtime)
- `random_seeds` = 1
- `cores` = 4
- `hyper_param_samples` = 8
- `bs` = 128 (in previous setup this was 64, the new 128 will reduce runtime)

**Comment by smilesun** [comment]
> @Car-la-F pointed out, that we could split the computation between different cluster rather easily, if we run the same configuration except different random seeds. By this, we can split the repetition for stochastic variance.

that is actually a good idea!
